File: python_target/routines/amount_utils.py
"""Converted from routines/AmountUtils.java.

Currency rounding, formatting, and amortization calculations.
"""
import locale
import math
from decimal import ROUND_HALF_UP, Decimal


def round_to_decimal(value: float, places: int) -> float:
    d = Decimal(str(value)).quantize(Decimal(10) ** -places, rounding=ROUND_HALF_UP)
    return float(d)


def format_currency(amount: float) -> str:
    try:
        locale.setlocale(locale.LC_ALL, "en_US.UTF-8")
    except locale.Error:
        pass
    return locale.currency(amount, grouping=True)


def calculate_monthly_payment(
    loan_amount: float, annual_rate: float, term_months: int
) -> float:
    if loan_amount <= 0 or term_months <= 0:
        return 0.0
    if annual_rate <= 0:
        return loan_amount / term_months
    monthly_rate = annual_rate / 100.0 / 12.0
    factor = math.pow(1 + monthly_rate, term_months)
    return round_to_decimal(loan_amount * (monthly_rate * factor) / (factor - 1), 2)


def is_valid_amount(amount: float, max_allowed: float) -> bool:
    return (
        amount > 0
        and amount <= max_allowed
        and not math.isnan(amount)
        and not math.isinf(amount)
    )


def cents_to_dollars(cents: int) -> float:
    return round_to_decimal(cents / 100.0, 2)

File: routines/AmountUtils.java
package routines;

import java.math.BigDecimal;
import java.math.RoundingMode;
import java.text.NumberFormat;
import java.util.Locale;

/**
 * Currency and amount utilities for financial ETL processing.
 * Used by: loan_application_ingest, loan_risk_scoring, payment_match,
 *          payment_discrepancy, payment_settlement, cfpb_extract, occ_compliance_report
 */
public class AmountUtils {

    /**
     * Round a double to the specified number of decimal places.
     * Uses HALF_UP rounding (standard financial rounding).
     */
    public static double roundToDecimal(double value, int places) {
        BigDecimal bd = BigDecimal.valueOf(value);
        bd = bd.setScale(places, RoundingMode.HALF_UP);
        return bd.doubleValue();
    }

    /**
     * Format a numeric amount as US currency string.
     * Example: 45000.5 → "$45,000.50"
     */
    public static String formatCurrency(double amount) {
        NumberFormat fmt = NumberFormat.getCurrencyInstance(Locale.US);
        return fmt.format(amount);
    }

    /**
     * Calculate monthly payment using standard amortization formula.
     * P = L * [r(1+r)^n] / [(1+r)^n - 1]
     * where L = loan amount, r = monthly rate, n = number of months
     */
    public static double calculateMonthlyPayment(double loanAmount, double annualRate, int termMonths) {
        if (loanAmount <= 0 || termMonths <= 0) return 0;
        if (annualRate <= 0) return loanAmount / termMonths;

        double monthlyRate = annualRate / 100.0 / 12.0;
        double factor = Math.pow(1 + monthlyRate, termMonths);
        return roundToDecimal(loanAmount * (monthlyRate * factor) / (factor - 1), 2);
    }

    /**
     * Validate that an amount is positive and within acceptable range.
     * Used for input validation before loading to Snowflake.
     */
    public static boolean isValidAmount(double amount, double maxAllowed) {
        return amount > 0 && amount <= maxAllowed && !Double.isNaN(amount) && !Double.isInfinite(amount);
    }

    /**
     * Convert cents to dollars.
     */
    public static double centsToDollars(long cents) {
        return roundToDecimal(cents / 100.0, 2);
    }
}
